xinference/model/llm/utils.py

```diff
@@ -535,26 +535,39 @@ def _eval_qwen_chat_arguments(c, tools):
             # Refer to:
             # https://github.com/QwenLM/Qwen/blob/main/examples/react_prompt.md
             # https://github.com/QwenLM/Qwen/blob/main/openai_api.py#L297
-            func_name, func_args = "", ""
+            func_name, func_args, content = "", "", ""
             i = text.rfind("\nAction:")
             j = text.rfind("\nAction Input:")
             k = text.rfind("\nObservation:")
+            t = max(
+                text.rfind("\nThought:", 0, i), text.rfind("Thought:", 0, i)
+            )  # find the last thought just before Action, considering the Thought at the very beginning
             if 0 <= i < j:  # If the text has `Action` and `Action input`,
                 if k < j:  # but does not contain `Observation`,
                     # then it is likely that `Observation` is omitted by the LLM,
                     # because the output text may have discarded the stop word.
                     text = text.rstrip() + "\nObservation:"  # Add it back.
                     k = text.rfind("\nObservation:")
-            if 0 <= i < j < k:
+            if 0 <= t < i < j < k:
                 func_name = text[i + len("\nAction:") : j].strip()
                 func_args = text[j + len("\nAction Input:") : k].strip()
+                content = text[
+                    t + len("\nThought:") : i
+                ].strip()  # len("\nThought:") and len("Thought:") both are OK since there is a space after :
             if func_name:
-                return None, func_name, json.loads(func_args)
-            z = text.rfind("\nFinal Answer: ")
-            if z >= 0:
-                text = text[z + len("\nFinal Answer: ") :]
+                return content, func_name, json.loads(func_args)
```

Comment:
Actually, as shown by the OpenAI API, the content should be null if the tool call is successful: https://platform.openai.com/docs/api-reference/chat/create
```
{
  "id": "chatcmpl-abc123",
  "object": "chat.completion",
  "created": 1699896916,
  "model": "gpt-3.5-turbo-0125",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "content": null,
        "tool_calls": [
          {
            "id": "call_abc123",
            "type": "function",
            "function": {
              "name": "get_current_weather",
              "arguments": "{\n\"location\": \"Boston, MA\"\n}"
            }
          }
        ]
      },
      "logprobs": null,
      "finish_reason": "tool_calls"
    }
  ],
  "usage": {
    "prompt_tokens": 82,
    "completion_tokens": 17,
    "total_tokens": 99
  }
}
```
@qinxuye I think we can merge this if it fixes some corner cases.

Reply:
OK, @codingl2k1 can you help to fix the corner case?

Reply:
I mean some corner use cases require the thought content, but OpenAI's API just returns null. If we strictly follow OpenAI's API, then we should return null if the tool call is successful.

Reply:
So, I think we can merge this PR even if it mismatches the OpenAI's API.

Reply:
OK, I will merge it